core/src/main/java/hudson/model/Queue.java

```diff
@@ -377,6 +396,9 @@ public synchronized void load() {
                             if(item.task==null)
                                 continue;   // botched persistence. throw this one away
 
+                            // backward compatibility - in case the serialized queue state was
+                            // a List of items (as of old), then the maxId is the max of the items in
+                            // the persisted queue.
```

Comment:
This seems convoluted.

If you are loading in a `List<Task>` from the old `queue.xml`, then we have to fall back to inferring a `maxId` as here, and setting `COUNTER` to it—fine.

But if we are loading a `State` from the new `queue.xml`, which is the normal code path going forward, we will ultimately set `COUNTER` to `state.counter`. Going through `maxId` in that case is just confusing.

IOW it would be clearer to have something like

```java
List<Task> tasks;
if (unmarshaledObj instanceof State) {
    tasks = ((State) unmarshaledObj).items;
    COUNTER = ((State) unmarshaledObj).counter;
} else {
    int maxId = 0;
    tasks = new ArrayList<Task>();
    for (Object o : (List) unmarshaledObj) {
        if (o instanceof Task) {
            tasks.add((Task) o);
            maxId = Math.max(maxId, ((Task) o).id);
        }
    }
    COUNTER = maxId;
}
```

which separates the compatibility code into its own block for easy ignoring.